ztag/stream.py

```diff
@@ -6,6 +6,76 @@
 from ztag.errors import IgnoreObject
 
 
+class Updater(object):
+    """
+    Updater encapsulates the behavior for the updates.csv file; put_update() is called with each
+    update, but output is only written every :frequency: seconds.
+    """
+    def __init__(self, output=None, frequency=1.0, logger=None):
+        self.output = output
+        self.frequency = frequency
+        self.logger = logger
+        self.prev = None
+        self._wrote_labels = False
+
+    def put_update(self, row):
+        if not self.output:
+            return
+
+        if self.prev and (row.time - self.prev.time) < self.frequency:
+            return
+
+        self.prev = row
+
+        if not self._wrote_labels:
+            self.output.write(row.get_csv_labels() + "\n")
+            self._wrote_labels = True
+
+        self.output.write(row.get_csv() + "\n")
+        self.output.flush()
+
+    def close(self):
+        if self.output and self.output != sys.stderr:
+            try:
+                self.output.close()
+            except BaseException as e:
+                if self.logger:
+                    self.logger.warn("Failed to close updates CSV stream: %s", str(e))
+
+
+class UpdateRow(object):
+    """
+    UpdateRow encapsulates the information for a single update and the logic for outputting it as
+    a CSV row.
+    """
+    ORDER = ("skipped", "handled", "delta_skipped", "delta_handled")
+
+    def __init__(self, skipped, handled, prev=None):
+        """
+        Construct a new row with the given number of skipped / handled entries, and calculate the
+        deltas from prev (or set them to 0). Also sets time to now.
+        :param skipped: current total number of skipped records
+        :param handled: current total number of handled records
+        :param prev: the previous UpdateRow
+        """
+        self.time = time.time()
```

Comment:
**Nit**: I'd allow the time optionally be passed in via a named argument, so that it's possible to create two identical objects / remove mutation from constructor.

Reply:
Good call. That will make the unit tests cleaner, too.